File: alerts-configurator/src/main/java/com/ocadotechnology/newrelic/alertsconfigurator/Configurator.java
package com.ocadotechnology.newrelic.alertsconfigurator;

import com.ocadotechnology.newrelic.alertsconfigurator.configuration.ApplicationConfiguration;
import com.ocadotechnology.newrelic.alertsconfigurator.configuration.PolicyConfiguration;
import com.ocadotechnology.newrelic.alertsconfigurator.internal.entities.EntityResolver;
import com.ocadotechnology.newrelic.api.NewRelicApi;
import lombok.NonNull;

import java.util.Collection;
import java.util.Collections;

public class Configurator {
    private final ApplicationConfigurator applicationConfigurator;
    private final PolicyConfigurator policyConfigurator;
    private final ConditionConfigurator conditionConfigurator;
    private final ExternalServiceConditionConfigurator externalServiceConditionConfigurator;
    private final ChannelConfigurator channelConfigurator;


    private Collection<ApplicationConfiguration> applicationConfigurations = Collections.emptyList();
    private Collection<PolicyConfiguration> policyConfigurations = Collections.emptyList();

    /**
     * NewRelic Alerts configurator constructor
     * @param apiKey API Key for given NewRelic account
     */
    public Configurator(@NonNull String apiKey) {
        NewRelicApi api = new NewRelicApi(apiKey);
        applicationConfigurator = new ApplicationConfigurator(api);
        policyConfigurator = new PolicyConfigurator(api);
        conditionConfigurator = new ConditionConfigurator(api, EntityResolver.defaultInstance());
        externalServiceConditionConfigurator = new ExternalServiceConditionConfigurator(api, EntityResolver.defaultInstance());
        channelConfigurator = new ChannelConfigurator(api);
    }

    Configurator(ApplicationConfigurator applicationConfigurator,
                 PolicyConfigurator policyConfigurator,
                 ConditionConfigurator conditionConfigurator,
                 ExternalServiceConditionConfigurator externalServiceConditionConfigurator,
                 ChannelConfigurator channelConfigurator) {
        this.applicationConfigurator = applicationConfigurator;
        this.policyConfigurator = policyConfigurator;
        this.conditionConfigurator = conditionConfigurator;
        this.externalServiceConditionConfigurator = externalServiceConditionConfigurator;
        this.channelConfigurator = channelConfigurator;
    }

    /**
     * Synchronizes configurations (see {@link Configurator#setApplicationConfigurations(Collection)}
     * and {@link Configurator#setPolicyConfigurations(Collection)}
     */
    public void sync() {
        for (ApplicationConfiguration applicationConfiguration : applicationConfigurations) {
            applicationConfigurator.sync(applicationConfiguration);
        }

        for (PolicyConfiguration configuration : policyConfigurations) {
            policyConfigurator.sync(configuration);
            conditionConfigurator.sync(configuration);
            externalServiceConditionConfigurator.sync(configuration);
            channelConfigurator.sync(configuration);
        }
    }

    /**
     * @param applicationConfigurations collection of application configurations to synchronize
     * @see ApplicationConfiguration
     */
    public void setApplicationConfigurations(@NonNull Collection<ApplicationConfiguration> applicationConfigurations) {
        this.applicationConfigurations = applicationConfigurations;
    }

    /**
     * @param policyConfigurations collection of policy configurations to synchronize
     * @see PolicyConfiguration
     */
    public void setPolicyConfigurations(@NonNull Collection<PolicyConfiguration> policyConfigurations) {
        this.policyConfigurations = policyConfigurations;
    }
}

File: newrelic-alerts-configurator-dsl/src/main/kotlin/com/ocadotechnology/newrelic/alertsconfigurator/dsl/Configurator.kt
package com.ocadotechnology.newrelic.alertsconfigurator.dsl

import com.ocadotechnology.newrelic.alertsconfigurator.Configurator
import com.ocadotechnology.newrelic.alertsconfigurator.configuration.ApplicationConfiguration
import com.ocadotechnology.newrelic.alertsconfigurator.configuration.PolicyConfiguration
import com.ocadotechnology.newrelic.alertsconfigurator.dsl.configuration.ApplicationConfigurations
import com.ocadotechnology.newrelic.alertsconfigurator.dsl.configuration.PolicyConfigurations

// TODO Add README.md - index
// TODO Add README.md - + operator
// TODO Add README.md - term and nrqlTerm flux
@NewRelicConfigurationMarker
class ConfiguratorDsl {
    internal val applicationConfigurations: MutableList<ApplicationConfiguration> = mutableListOf()
    internal val policyConfigurations: MutableList<PolicyConfiguration> = mutableListOf()

    fun applications(block: ApplicationConfigurations.() -> Unit) = applicationConfigurations.addAll(ApplicationConfigurations().apply(block).applications)
    fun policies(block: PolicyConfigurations.() -> Unit) = policyConfigurations.addAll(PolicyConfigurations().apply(block).policies)
}

fun configurator(apiKey: String, block: ConfiguratorDsl.() -> Unit) {
    val dsl = ConfiguratorDsl()
    dsl.apply(block)

    val configurator = Configurator(apiKey)
    configurator.setApplicationConfigurations(dsl.applicationConfigurations)
    configurator.setPolicyConfigurations(dsl.policyConfigurations)
    configurator.sync()
}